### ISSUE TITLE
Realm create running out of memory and crashing

### ISSUE THREAD

**Issue by Genhain** [realm/realm-js#2536]
## Goals

download and insert 19000+ records into the database paginated by 200 per request

## Expected Results

Insert 19000+ records into database

## Actual Results

After about 9000 records, the app crashes, and if i check XCode, it says terminated due to memory issues

![image](https://user-images.githubusercontent.com/2082208/65524502-78ca9280-df20-11e9-9704-43ae855bcd1e.png)

There are no leaks, that i know of, i just leave it to download and eventually it crashes.
Another thing to note is that each time `create` is called it gradually get's longer and longer every 200 records. it starts at about 0.8s for 200 then right before it crashes it get's up to about 9.0s, i once deleted most of the UI elements to purely let it download and it got up to 14000 records total, and the last 200 it tried to insert took 50.0+ seconds. 

Another curiosity is that upon booting the project back up, it loaded those 14000 records in barely anytime at all and the memory consumption barely moved. My current findings seems to be that calling create periodically is causing the issue. 

14000 recrods over 70 fetch requests.

## Steps to Reproduce

TBD I don't have a surefire way as of yet, i will try and create a RN project and emulate it as best i can to see if i can get it happen in a more isolated environment.

## Code Sample

currently i have a singleton `realm.js` file that should be instantiated once it can be found [here](https://gist.github.com/Genhain/aada8b6d80bef0cb971140239543cf98)

My api call uses RXJS to do paginated calls like so
```ts
static fetchAllAttendees = (path: string, authToken: string) => {
    return from(AttendeeFetcher._fetchAllAttendees(path, 1, authToken)).pipe(
      expand(({ meta }) => {
        return meta.pagination_info.next_page
          ? AttendeeFetcher._fetchAllAttendees(
              path,
              meta.pagination_info.next_page,
              authToken
            )
          : empty()
      }),
      map((x: any) => plainToClass(Attendee, <any[]>x.attendees)),
      catchError(err => throwError(err))
    )
  }
```
and i use `class-transformer` lib to transform plain JSON into a typed class with `plainToClass`

and here is where i subscribe to get the attendees

```ts
networkRouter
  .get(NetworkResource.Attendee, NetworkPath.getPaginated, {
    id: this.party.id,
    authToken: user.authToken
  })
  .subscribe({
    next: attendees => {
      reactotron.logImportant!(attendees)
      let bench = reactotron.benchmark!("attendee write")
      EXRealm.shared.write(() => {
        let i = 0
        let length = attendees.length
        for(; i< length; i++) {
          EXRealm.shared.create(Attendee.schema.name, attendees[i], true)
        }
        bench.step("created/updated attendees")
      })
      bench.stop("attendee write done")
    },
    error: reactotron.logImportant!
  })
```
i also create an observable out of the realm listener to response to updates
```ts
this.attendeeListener = EXRealm.shared
  .objects<Attendee>(Attendee.schema.name)
  .filtered(`partyId = ${this.party.id}`)
  .sorted("name", false)

this.realmAttendees$ = new Observable<AttendeeChanges>(observer => {
  const handler = (
    attendees: Collection<Attendee & Realm.Object>,
    changes: CollectionChangeSet
  ) => observer.next([attendees,changes])
  
  this.attendeeListener.addListener(handler)
  return () => this.attendeeListener.removeListener(handler)
})
```
## Version of Realm and Tooling

- Realm JS SDK Version: 3.1.0
- Node or React Native: React Native
- Client OS & Version: iOS 12
- Which debugger for React Native: Reactotron

Im not certain if related or not but i did come across #2499 
I will continue to tinker with it to see if it's perhaps something I'm doing wrong and will update accordingly, but thought i should at least get this posted to see if anyone has any insight?

**Comment by Genhain** [comment]
#2499

**Comment by Genhain** [comment]
## Update
Still using up a lot of memory but i have fixed the exponentially growing write times.
i removed all object relationships and only save primitives and it fixes that particular issue.

[here is the model](https://gist.github.com/Genhain/feea260fae151a3c4d8897878c98f25a)
i removed the relationships `answers` and `customLines`

**Comment by thekevinbrown** [comment]
@Genhain if you're expecting data that you've written to be queryable from your listener, then #2499 is related. I show in the repro on that issue that newly created objects are not returned in the listener, which is supposedly a feature. Not sure how to leverage that feature without further guidance over on that issue, but if I were you I'd try replacing Realm listeners with your own event emitter setup to see if it's any better.

**Comment by Genhain** [comment]
@thekevinbrown If i'm understanding this correctly then it seems we might have different behaviour, because the listener does contain the newly created/inserted objects.

my current issue It is just that over the course of creating/inserting paginated data it begins using up a significant amount of memory. perhaps because `.write` is called every time a new page comes in and i recall doing this often enough leads to poor performance.

as for rolling your own event emitters and my assumption is that to create your own event emitters, you periodically query the database and then emit the delta?

**Comment by thekevinbrown** [comment]
Fair enough!

I'm just making sure writes happen through a controller object which also emits events to listeners after the write call completes.

**Comment by kneth** [comment]
@Genhain I have no real experience with RxJS so I no useful information to share out the interaction between Realm and RxJS.

It could be interesting to see what happens if you call `Realm.compact()` just after the `write()` call. It could also be interesting to know how large the Realm file is after 1000 or 10000 inserts.

**Comment by Genhain** [comment]
@kneth I also changed from RXJS to just a recursive function, my initial worry about why the write times where getting larger per iteration was because React Native JS (or in my case TS) is not tail call optimised. But as I said the growing write times were solved when I removed the relationships from the schema.

> It could also be interesting to know how large the Realm file is after 1000 or 10000 inserts.

I never actually checked the realm file size itself, i was just checking the memory usage of the device. The memory usage still grows and does not shrink, But until more tests are done i suspect that is more my own doing than Realm, potentially both. Either way ill try calling `Realm.compact()` and let you know what i find.
